gibbon-rts/Makefile

```diff
@@ -137,8 +137,8 @@ $(BUILD_DIR)/%.o: $(C_RTS_DIR)/%.o
 	mkdir -p $(BUILD_DIR) && \
 	mv $^ $@
 
-$(C_RTS_DIR)/%.o: $(C_RTS_DIR)/%.c
-	$(CC) $(CFLAGS) -c -o $@ $^
+$(C_RTS_DIR)/(NAME).o:
+	$(CC) $(CFLAGS) -c -o $@ $(C_RTS_DIR)/(NAME).c
```

Comment:
Oh, excellent point! I'll add a comment.

What do you think about the Rust part? I didn't analyze it for the same problem but perhaps it suffers from the same...

Reply:
Ah, right, the Rust target will have to change as well. I was thinking that that should be OK since `cargo` factors in the flags passed to it when deciding whether to rebuild or not, but, forgot that we still need to unconditionally invoke `cargo build` every time.

I did some more googling and found [force targets](https://www.gnu.org/software/make/manual/make.html#Force-Targets). It looks like the effect is somewhat similar to marking a target "phony", but "FORCE" will specify the intent better perhaps? I tested this patch locally and it seems to be working; make will now rebuild the C and Rust code every time.

```diff
diff --git a/gibbon-rts/Makefile b/gibbon-rts/Makefile
index a26b0716..133656cb 100644
--- a/gibbon-rts/Makefile
+++ b/gibbon-rts/Makefile
@@ -130,7 +130,9 @@ $(BUILD_DIR)/$(RUST_RTS_SO): $(RUST_RTS_PATH)
        mkdir -p $(BUILD_DIR) && \
        mv $^ $@

-$(RUST_RTS_PATH): $(RUST_SOURCES)
+# The Rust sources are not specified as prerequisites and a FORCE
+# prerequisite is used to force unconditional recompilation.
+$(RUST_RTS_PATH): FORCE
        cd  $(RUST_RTS_DIR) && \
        $(RSC) build -p gibbon-rts-ng $(RSFLAGS)

@@ -141,8 +143,10 @@ $(BUILD_DIR)/%.o: $(C_RTS_DIR)/%.o
        mkdir -p $(BUILD_DIR) && \
        mv $^ $@

-$(C_RTS_DIR)/%.o: $(C_RTS_DIR)/%.c
-       $(CC) $(CFLAGS) -c -o $@ $^
+# The C sources are not specified as prerequisites and a FORCE
+# prerequisite is used to force unconditional recompilation.
+$(C_RTS_DIR)/$(NAME).o: FORCE
+       $(CC) $(CFLAGS) -c -o $@ $(C_RTS_DIR)/$(NAME).c

 $(BUILD_DIR)/%.h: $(C_RTS_DIR)/%.h
        mkdir -p $(BUILD_DIR) && \
@@ -154,4 +158,7 @@ $(BUILD_DIR):
 clean:
        rm -rf $(BUILD_DIR) $(RUST_RTS_DIR)/target

+# See https://www.gnu.org/software/make/manual/make.html#Force-Targets
+FORCE:
+
 .PHONY: clean
```

The second FORCE isn't required and your change works by itself, but I kept it just to make it explicit. Wrt the `$(RUST_RTS_SO_PATH)` change, another option is to mark this target "phony" but I didn't do that. I don't feel strongly either way though.

What do you think?

Reply:
Added a comment.

Reply:
Oh wow, what a timing! Your approach looks much more principled. Could you drop your version of the Makefile instead of a patch here? That would save me a minute creating a patch file locally and remembering git incantations to apply it :-)

Reply:
Sounds good, will do.

Reply:
I see https://github.com/iu-parfunc/gibbon/pull/245/commits/02a014cfc2c9d899147f0a203ecb6206657dc054 -- thank you!!